lib/ecto/query/builder/limit_offset.ex

```diff
@@ -5,39 +5,105 @@ defmodule Ecto.Query.Builder.LimitOffset do
   """
-  @spec build(:limit | :offset, Macro.t, [Macro.t], Macro.t, Macro.Env.t) :: Macro.t
+  @spec build(:limit | :with_ties | :offset, Macro.t(), [Macro.t()], Macro.t(), Macro.Env.t()) ::
+          Macro.t()
   def build(type, query, binding, expr, env) do
-    {query, binding} = Builder.escape_binding(query, binding, env)
-    {expr, {params, _acc}} = Builder.escape(expr, :integer, {[], %{}}, binding, env)
+    {query, vars} = Builder.escape_binding(query, binding, env)
+    {expr, {params, _acc}} = escape(type, expr, {[], %{}}, vars, env)
     params = Builder.escape_params(params)
+    quoted = build_quoted(type, expr, params, env)
+
+    Builder.apply_query(query, __MODULE__, [type, quoted], env)
+  end
+
+  defp escape(type, expr, params_acc, vars, env) when type in [:limit, :offset] do
+    Builder.escape(expr, :integer, params_acc, vars, env)
+  end
+
+  defp escape(:with_ties, expr, params_acc, _vars, _env) when is_boolean(expr) do
+    {expr, params_acc}
+  end
+
+  defp escape(:with_ties, {:^, _, [expr]}, params_acc, _vars, _env) do
+    {quote(do: Ecto.Query.Builder.LimitOffset.with_ties!(unquote(expr))), params_acc}
+  end
+
+  defp escape(:with_ties, expr, _params_acc, _vars, _env) do
+    Builder.error!(
+      "`with_ties` expression must be a compile time boolean or an interpolated value using ^, got: `#{Macro.to_string(expr)}`"
+    )
+  end
 
-    limoff = quote do: %Ecto.Query.QueryExpr{
-                        expr: unquote(expr),
-                        params: unquote(params),
-                        file: unquote(env.file),
-                        line: unquote(env.line)}
+  defp build_quoted(:limit, expr, params, env) do
+    quote do: %Ecto.Query.LimitExpr{
+            expr: unquote(expr),
+            params: unquote(params),
+            with_ties: false,
+            file: unquote(env.file),
+            line: unquote(env.line)
+          }
+  end
 
-    Builder.apply_query(query, __MODULE__, [type, limoff], env)
+  defp build_quoted(:offset, expr, params, env) do
+    quote do: %Ecto.Query.QueryExpr{
+            expr: unquote(expr),
+            params: unquote(params),
+            file: unquote(env.file),
+            line: unquote(env.line)
+          }
   end
 
+  defp build_quoted(:with_ties, expr, _params, _env), do: expr
+
   @doc """
   The callback applied by `build/4` to build the query.
   """
-  @spec apply(Ecto.Queryable.t, :limit | :offset, term) :: Ecto.Query.t
+  @spec apply(Ecto.Queryable.t(), :limit | :with_ties | :offset, term) :: Ecto.Query.t()
   def apply(%Ecto.Query{} = query, :limit, expr) do
     %{query | limit: expr}
   end
+
+  def apply(%Ecto.Query{limit: limit} = query, :with_ties, expr) do
+    %{query | limit: apply_limit(limit, expr)}
+  end
+
   def apply(%Ecto.Query{} = query, :offset, expr) do
     %{query | offset: expr}
   end
+
   def apply(query, kind, expr) do
     apply(Ecto.Queryable.to_query(query), kind, expr)
   end
+
+  def apply_limit(nil, _with_ties) do
```

Comment:
Should this function be `defp`?

Reply:
Oh I made it public because the builder is unescaping the query struct but not the nested structs. So had to [do this](https://github.com/greg-rychlewski/ecto/blob/with_ties/lib/ecto/query/builder/limit_offset.ex#L104). There is something [similar here with ctes](https://github.com/greg-rychlewski/ecto/blob/with_ties/lib/ecto/query/builder/cte.ex#L82).

Does it seem ok? This part tripped me up the most not 100% sure I did it the best way.

Reply:
Doh, I saw just now that it has been called from the quote too. So this is good, just add a small `@doc` to align with the rest of the module?

Reply:
done!